File: scripts/tf_latency.py
#!/usr/bin/env python

import rospy
from tf import TransformListener

class TFLatency:
    def __init__(self, source_frame, target_frame):
        self.source_frame = source_frame
        self.target_frame = target_frame
        self.listener = TransformListener(1, rospy.Duration(0.01))

    def run(self):
        self.listener.waitForTransform(self.source_frame, self.target_frame, rospy.Time(), rospy.Duration(5.0))
        sum = 0
        count = 0
        while not rospy.is_shutdown():
            now = rospy.Time.now()
            if self.listener.canTransform(self.source_frame, self.target_frame, rospy.Time(0)):
                t = self.listener.getLatestCommonTime(self.source_frame, self.target_frame)
                delta = (now - t).to_sec() * 1000 #ms
                sum += delta
                count +=1
                rospy.loginfo("Latency: %f ms (AVG: %f ms)", delta, sum / count)
                #if delta > 20:
                #    self.listener.clear()
            rospy.sleep(0.01)

if __name__ == '__main__':

    rospy.init_node('tf_latency', anonymous=True)
    source_frame = rospy.get_param("~source_frame", "/world")
    target_frame = rospy.get_param("~target_frame")
    tflatency = TFLatency(source_frame, target_frame)
    tflatency.run()

File: src/tf_latency.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>

int main(int argc, char** argv) {
  ros::init(argc, argv, "tf_latency");
  ros::NodeHandle node("~");

  std::string source_frame = "/world";
  std::string target_frame;
  node.getParam("target_frame", target_frame);

  // Reducing the cache size from 10 seconds to any other value doesn't seem to change anything
  tf::TransformListener listener(ros::Duration(0.02));

  listener.waitForTransform(source_frame, target_frame, ros::Time(0), ros::Duration(5.0));
  float sum = 0;
  int count = 0;

  while (node.ok()) {
    ros::Time now = ros::Time::now();
    ros::Time t;
    listener.getLatestCommonTime(source_frame, target_frame, t, NULL);
    float delta = (now - t).toSec() * 1000; //ms
    sum += delta;
    count +=1;
    ROS_INFO("Latency: %f ms (AVG: %f ms)", delta, sum / count);

    // Clearing the buffer reduces the latency, however can cause that there is no transformation available in the next iteration
    //listener.clear();

    ros::Duration(0.01).sleep();
  }

  return 0;
}
